<launch>
  <arg name="DEVICE"/>
  <arg name="FPS" default="15/1"/>
  <arg name="width"/>
  <arg name="height"/>
  <arg name="camera_info_file_right" default="" />
  <arg name="camera_info_file_left" default="" />
  <arg name="camera_info_url_default" default="" />

  <arg name="left_offset"/>
  <arg name="left_width"/>
  <arg name="right_offset"/>
  <arg name="right_width"/>


  <arg name="camera_info_url_right" default="file://$(arg camera_info_file_right)" />
  <arg name="camera_info_url_left" default="file://$(arg camera_info_file_left)" />

  <arg name="manager" default="manager" /> <!-- Must be globally qualified -->

  <node pkg="nodelet" type="nodelet" name="$(arg manager)" args="manager" output="screen" />

  <node name="gscam_driver" pkg="nodelet" type="nodelet" args="load gscam/GSCamNodelet $(arg manager)" output="screen">
    <param name="camera_name" value="default"/>
    <param name="gscam_config" value="v4l2src device=$(arg DEVICE) ! video/x-raw-yuv,framerate=$(arg FPS),width=$(arg width),height=$(arg height) ! ffmpegcolorspace"/>
    <param name="frame_id" value="/CameraStereo_optical_frame"/>
    <param name="sync_sink" value="true"/>
    <param name="camera_info_url" value="$(arg camera_info_url_default)" />
  </node>

  <!-- Split image to left and right image -->
  <node pkg="nodelet" type="nodelet" name="split_right" args="load image_proc/crop_decimate $(arg manager)">
    <param name="queue_size" type="int" value="10" />
    <param name="decimation_x" type="int" value="1" />
    <param name="decimation_y" type="int" value="1" />
    <param name="x_offset" type="int" value="$(arg right_offset)" />
    <param name="y_offset" type="int" value="0" />
    <param name="width" type="int" value="$(arg right_width)" />
    <param name="height" type="int" value="$(arg height)" />
    <remap from="camera_out/image_raw" to="right/image_raw" />
    <!-- Throw original camera info -->
    <remap from="right/camera_info" to="/null/right/camera_info" />
  </node>

  <node pkg="nodelet" type="nodelet" name="split_left" args="load image_proc/crop_decimate $(arg manager)">
    <param name="queue_size" type="int" value="10" />
    <param name="decimation_x" type="int" value="1" />
    <param name="decimation_y" type="int" value="1" />
    <param name="x_offset" type="int" value="$(arg left_offset)" />
    <param name="y_offset" type="int" value="0" />
    <param name="width" type="int" value="$(arg left_width)" />
    <param name="height" type="int" value="$(arg height)" />
    <remap from="camera_out/image_raw" to="left/image_raw" />
    <!-- Throw original camera info -->
    <remap from="left/camera_info" to="/null/left/camera_info" />
  </node>

  <node pkg="stereo_driver" type="camera_info_publisher.py" name="camera_info_publisher" >
    <param name="file_name" value="$(arg camera_info_file_left)" />
    <param name="camera_name" value="/camera" />
  </node>

  <node pkg="stereo_driver" type="camera_info_publisher.py" name="camera_left_info_publisher" >
    <param name="file_name" value="$(arg camera_info_file_left)" />
    <param name="camera_name" value="left" />
  </node>

  <node pkg="stereo_driver" type="camera_info_publisher.py" name="camera_right_info_publisher" >
    <param name="file_name" value="$(arg camera_info_file_right)" />
    <param name="camera_name" value="right" />
  </node>
</launch>
